### ISSUE TITLE
demo: https://codepen.io/iangilman/pen/qqQLPY doesn't work on mobile

### ISSUE THREAD

**Issue by zmomozq** [openseadragon/openseadragon#2398]
Hi, I used your case: https://codepen.io/iangilman/pen/qqQLPY. It works the same as yours on a computer browser, but why doesn’t it work well on mobile devices? For example, when dragging quickly on a mobile device, I clearly called fitBounds, but it didn't seem to work.

**Comment by zmomozq** [comment]
Another question is, can the initialization parameter collectionTileMargin be set dynamically?

**Comment by iangilman** [comment]
Interesting... What I'm seeing on mobile (iPhone Safari) is it doesn't snap to position after you let up dragging. Is that what you're seeing? Have you tested to see if the `canvas-drag-end` is firing on mobile? @msalsbery any thoughts?

For your second question, we don't have a setter for `collectionTileMargin` but you can change it. You can set it directly on your viewer like `viewer.collectionTileMargin = 20`, but it won't take effect until you add another image. Another option is to call `viewer.world.range` directly using whatever values you want:

https://openseadragon.github.io/docs/OpenSeadragon.World.html#arrange

**Comment by zmomozq** [comment]
> 
Thank you for your reply. Yes, when you stop dragging, the viewport will not move to the specified position, but there is a way, that is to use the animation-finish event to monitor the end of the animation, and set it after the animation ends. Place the viewport at the specified position, but this is too slow. I don't know if it is because the viewport is still executing animation, so the position of the viewport cannot be set, but this problem does not exist on web browsers

**Comment by zmomozq** [comment]
Yes. I fire it on the canvas-drag-end event and print the border closest to where it fired. I did find the nearest borders and set them, and it was set, but it wasn't responding on the mobile device. Is it because finger triggered events are different from web browser mouse events? I tried to set it after animation-finish, no problem, but it was too slow to wait for the viewport movement animation to finish.

**Comment by iangilman** [comment]
Thank you for confirming. I would say that's definitely a bug, and yes, having to do with the different events used internally between mouse and touch.

Have you tried `canvas-release`?

**Comment by zmomozq** [comment]
Yes, I've tried them all and they work fine in a web browser, but it doesn't work when switching to a mobile emulator or using a real mobile device

**Comment by zmomozq** [comment]
> Thank you for confirming. I would say that's definitely a bug, and yes, having to do with the different events used internally between mouse and touch.
> 
> Have you tried `canvas-release`?

Yes, I've tried them all and they work fine in a web browser, but it doesn't work when switching to a mobile emulator or using a real mobile device

**Comment by zmomozq** [comment]
> Can you confirm that the `canvas-drag-end` isn't firing on mobile, by putting a console.log in there or something? If so, it sounds like a bug. If you're interested in pursuing it, I can help point you in the right direction. @msalsbery might have some ideas as well.
> 
> Otherwise, another thing to try would be the `canvas-release` event... see if it fires on mobile.

I answered this question incorrectly, canvas-drag-end is triggered, but the fitBounds call is not triggered when canvas-drag-end is triggered

**Comment by iangilman** [comment]
I see, thank you for clarifying! I guess it's a matter of debugging the code then… in the `canvas-drag-end` handler in https://codepen.io/iangilman/pen/qqQLPY, there are three `if` statements… presumably one of those is failing a test. You can drop some `console.logs` in there to see which one it is.

**Comment by zmomozq** [comment]
> I see, thank you for clarifying! I guess it's a matter of debugging the code then… in the `canvas-drag-end` handler in https://codepen.io/iangilman/pen/qqQLPY, there are three `if` statements… presumably one of those is failing a test. You can drop some `console.logs` in there to see which one it is.

It's not that one of the three ifs isn't satisfied, I've already used console.log to output the bounds that need to be set, and set them. But it doesn't work on mobile devices. I hope you can try a quick drag and drop on mobile devices. It doesn't work when you call fitBounds

**Comment by zmomozq** [comment]
> Interesting... What I'm seeing on mobile (iPhone Safari) is it doesn't snap to position after you let up dragging. Is that what you're seeing? Have you tested to see if the `canvas-drag-end` is firing on mobile? @msalsbery any thoughts?
> 
> For your second question, we don't have a setter for `collectionTileMargin` but you can change it. You can set it directly on your viewer like `viewer.collectionTileMargin = 20`, but it won't take effect until you add another image. Another option is to call `viewer.world.range` directly using whatever values you want:
> 
> https://openseadragon.github.io/docs/OpenSeadragon.World.html#arrange

As you said in response to my first question

**Comment by msalsbery** [comment]
At first glance it looks like a behavior difference depending on if the drag ends with a flick gesture or not.

Can you confirm it works if you stop drag movement then remove touch finger?

**Comment by zmomozq** [comment]
> 

Removing the finger after stopping the drag calls to `fitBounds` work during slower drags. I marked the end of the drag in the `canvas-drag-end` event, checked the presence of the end of the drag in the `animation-finish` event, and then called the `fitBounds` to set the coordinates of the viewport. It was okay, but it took too long to wait

**Comment by zmomozq** [comment]
> At first glance it looks like a behavior difference depending on if the drag ends with a flick gesture or not.
> 
> Can you confirm it works if you stop drag movement then remove touch finger?

Removing the finger after stopping the drag calls to `fitBounds` work during slower drags. I marked the end of the drag in the `canvas-drag-end` event, checked the presence of the end of the drag in the `animation-finish` event, and then called the `fitBounds` to set the coordinates of the viewport. It was okay, but it took too long to wait

**Comment by zmomozq** [comment]
`            let dragEnd = false;
            viewer.addHandler("canvas-drag-end", () => {
                // Suppose I want to set the viewport to the viewport coordinates of the first image
                // On mobile devices, calling fitBounds here has no effect
                let curBounds = viewer.world.getItemAt(0).getBounds();
                viewer.viewport.fitBounds(curBounds);

                dragEnd = true;
            })

            viewer.addHandler("animation-finish", () => {
                if (dragEnd) {
                    // The call to fitBounds is valid at this point
                    let curBounds = viewer.world.getItemAt(0).getBounds();
                    viewer.viewport.fitBounds(curBounds);

                    dragEnd = false;
                }
            })`

**Comment by zmomozq** [comment]
```
            let dragEnd = false;
            viewer.addHandler("canvas-drag-end", () => {
                // Suppose I want to set the viewport to the viewport coordinates of the first image
                // On mobile devices, calling fitBounds here has no effect
                let curBounds = viewer.world.getItemAt(0).getBounds();
                viewer.viewport.fitBounds(curBounds);

                dragEnd = true;
            })

            viewer.addHandler("animation-finish", () => {
                if (dragEnd) {
                    // The call to fitBounds is valid at this point
                    let curBounds = viewer.world.getItemAt(0).getBounds();
                    viewer.viewport.fitBounds(curBounds);

                    dragEnd = false;
                }
            })
```

**Comment by msalsbery** [comment]
> 
> 
> Removing the finger after stopping the drag calls to `fitBounds` work during slower drags. I marked the end of the drag in the `canvas-drag-end` event, checked the presence of the end of the drag in the `animation-finish` event, and then called the `fitBounds` to set the coordinates of the viewport. It was okay, but it took too long to wait

[Here's a fork of Ian's CodePen.](https://codepen.io/msalsbery/pen/PoXpmPG) In the 'canvas-drag-end' event handler I set the event object's `preventDefaultAction` property to true. That will disable the viewer's processing of the flick gesture for all pointer devices that have that gesture enabled (just touch devices by default). That should make it behave the same as using a mouse. Note I added the setting of preventDefaultAction in the ` if (intersectionCount > 1) {` block...you may like the behavior better if you set the flag true outside that block.

You said waiting for the end of the pan animation to call fitBounds took too long...you could try setting the fitBounds 'immediately' parameter to true. That may help as it'll prevent pan animation for that call.

I'm a fan of using the flick gesture on touch devices, so I personally would make a more robust solution by using the flick speed to calculate where the pan is going to end [like OpenSeadragon.Viewer does](https://openseadragon.github.io/docs/openseadragon.js.html#line10777) and use those end position bounds to calculate the stuff for fitBounds. That would make it possible to flick the strip and have it settle to the correct place more smoothly.

**Comment by zmomozq** [comment]
> > Removing the finger after stopping the drag calls to `fitBounds` work during slower drags. I marked the end of the drag in the `canvas-drag-end` event, checked the presence of the end of the drag in the `animation-finish` event, and then called the `fitBounds` to set the coordinates of the viewport. It was okay, but it took too long to wait
> 
> [Here's a fork of Ian's CodePen.](https://codepen.io/msalsbery/pen/PoXpmPG) In the 'canvas-drag-end' event handler I set the event object's `preventDefaultAction` property to true. That will disable the viewer's processing of the flick gesture for all pointer devices that have that gesture enabled (just touch devices by default). That should make it behave the same as using a mouse. Note I added the setting of preventDefaultAction in the ` if (intersectionCount > 1) {` block...you may like the behavior better if you set the flag true outside that block.
> 
> You said waiting for the end of the pan animation to call fitBounds took too long...you could try setting the fitBounds 'immediately' parameter to true. That may help as it'll prevent pan animation for that call.
> 
> I'm a fan of using the flick gesture on touch devices, so I personally would make a more robust solution by using the flick speed to calculate where the pan is going to end [like OpenSeadragon.Viewer does](https://openseadragon.github.io/docs/openseadragon.js.html#line10777) and use those end position bounds to calculate the stuff for fitBounds. That would make it possible to flick the strip and have it settle to the correct place more smoothly.

Cool.  Thanks for your advice, I didn't expect that the failure to call `fitBounds` in the `canvas-drag-end` event was caused by other default events of the pointer being triggered at the end of the drag.  According to your instructions, I have already done this in `canvas-drag-end` event. `preventDefaultAction = true` in the end event prevents other default events from being triggered by the pointer, and the effect is very good.  I will also try to use the flick speed to calculate the end position of the translation based on your suggestion, thank you

**Comment by msalsbery** [comment]
> 
> I will also try using flick speed to calculate where the pan ends as you suggested, thank you

Right on. It should be relatively simple if you mimic what OpenSeadragon.Viewer does

**Comment by zmomozq** [comment]
> > I will also try using flick speed to calculate where the pan ends as you suggested, thank you
> 
> Right on. It should be relatively simple if you mimic [what OpenSeadragon.Viewer.onCanvasDragEnd does](https://openseadragon.github.io/docs/openseadragon.js.html#line10777)

okay, thank you
